### DISCUSSION TITLE
Move controls to another axaml file with seperate viewmodel

### DISCUSSION THREAD

**Discussion by Shadowblitz16** [AvaloniaUI/Avalonia#16815] [Q&A]
Hello I recently got rider again and I tried to move my controls into a seperate xaml file with a seperate viewmodel
but I got layout bugs when trying to use usercontrol so I just swapped the usercontrol root out for the root element type like so...

```xaml
<ScrollViewer xmlns="https://github.com/avaloniaui"
              xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
              xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
              xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
              xmlns:am="using:AvaloinaMario"
              xmlns:paz="using:Avalonia.Controls.PanAndZoom"
              mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
              x:Class="AvaloinaMario.EditorCanvasView"
              x:DataType="am:EditorCanvasViewModel"
              HorizontalScrollBarVisibility="Hidden" 
              VerticalScrollBarVisibility="Hidden">
    <paz:ZoomBorder MinZoomX="{Binding MinZoom}" MinZoomY="{Binding MinZoom}" MaxZoomX="{Binding MaxZoom}" MaxZoomY="{Binding MaxZoom}" EnableConstrains="True">
        <Canvas Background="{Binding Background}" Width="{Binding Width}" Height="{Binding Height}">
                
        </Canvas>
    </paz:ZoomBorder>
</ScrollViewer>

```

but I got these errors when compiling...
```bash
0>EditorCanvasView.axaml(1,2): Error AVLN2000 Avalonia: Unable to resolve suitable regular or attached property HorizontalScrollBarVisibility on type AvaloinaMario:AvaloinaMario.EditorCanvasView Line 1, position 2.
0>EditorCanvasView.axaml(1,2): Error AVLN2000 Avalonia: Unable to resolve suitable regular or attached property VerticalScrollBarVisibility on type AvaloinaMario:AvaloinaMario.EditorCanvasView Line 1, position 2.
0>EditorMenuView.axaml(1,2): Error AVLN3000 Avalonia: Unable to find a setter that allows multiple assignments to the property Content of type Avalonia.Controls:Avalonia.Controls.ContentControl Line 1, position 2.
0>EditorEditMenuView.axaml(1,2): Error AVLN2000 Avalonia: Unable to resolve suitable regular or attached property Header on type AvaloinaMario:AvaloinaMario.EditorEditMenuView Line 1, position 2.
0>EditorEditMenuView.axaml(1,2): Error AVLN3000 Avalonia: Unable to find a setter that allows multiple assignments to the property Content of type Avalonia.Controls:Avalonia.Controls.ContentControl Line 1, position 2.
0>EditorFileMenuView.axaml(1,2): Error AVLN2000 Avalonia: Unable to resolve suitable regular or attached property Header on type AvaloinaMario:AvaloinaMario.EditorFileMenuView Line 1, position 2.
0>EditorFileMenuView.axaml(1,2): Error AVLN3000 Avalonia: Unable to find a setter that allows multiple assignments to the property Content of type Avalonia.Controls:Avalonia.Controls.ContentControl Line 1, position 2.
0>EditorHelpMenuView.axaml(1,2): Error AVLN2000 Avalonia: Unable to resolve suitable regular or attached property Header on type AvaloinaMario:AvaloinaMario.EditorHelpMenuView Line 1, position 2.
0>EditorToolsMenuView.axaml(1,2): Error AVLN2000 Avalonia: Unable to resolve suitable regular or attached property Header on type AvaloinaMario:AvaloinaMario.EditorToolsMenuView Line 1, position 2.
0>EditorToolsMenuView.axaml(1,2): Error AVLN3000 Avalonia: Unable to find a setter that allows multiple assignments to the property Content of type Avalonia.Controls:Avalonia.Controls.ContentControl Line 1, position 2.
```

**Comment by stevemonaco** [comment]
I wouldn't recommend deriving from any `TemplatedControl` which has a defined `ControlTheme` and trying to use XAML like this. What value do you believe you're getting instead of sticking with a `UserControl`?

**Reply by Shadowblitz16** [reply]
User control breaks the layout when using it in menus and menu items.
I think the issue is that the root of the control is treated as a user control rather then something like a menuitem.

**Reply by stevemonaco** [reply]
A repro here might help someone give better advice. `MenuItem.Header` is an `object?` that you can put whatever content you want in, including a `UserControl`. In the Fluent `ControlTheme`, it gets presented via `ContentPresenter` and allows you to use a `HeaderTemplate` if you'd want:

https://github.com/AvaloniaUI/Avalonia/blob/de86c5a058915fca10adae4fc66b3e3ac8a8df86/src/Avalonia.Themes.Fluent/Controls/MenuItem.xaml#L98-L104

I do have to say that putting a `ScrollViewer` as the Header's direct child would seem odd.

**Reply by Shadowblitz16** [reply]
@stevemonaco I am just trying to separate my ui components out into files without breaking the layout.

I posted a bad example but I am also doing this with my Menu and MenuItems too.
The problem is, at least for menus is a UserControl breaks the layout.

Then the Menu thinks I am passing a UserControl as a menu item rather then a MenuItem because the MenuItem is a Child of the UserControl.

That's why I want to inherit the type of Control as the xml root directly instead of just a UserControl, so I don't have weird layout issues.

Basicly what I want to do is..
If I have something like this...

MenuView.xaml
```xml
<UserControl
...>
  <MenuFileNewView/>
</UserControl>
```

MenuFileNewView.xaml...
```xml

<UserControl
...>
    <MenuItem Header="_New"/>
</UserControl>
```

I actually want something like this...
MenuView.xaml
```xml
<Menu
...>
    <MenuFileNewView/>
</Menu>
```

MenuFileNewView.xaml...
```xml

<MenuItem
Header="_New"
...>
</MenuItem>
```

so the layout doesn't get all messed up.

**Reply by stevemonaco** [reply]
This is an antipattern that I call over-componentization. You try to create a lot of small components like you're writing React (you're not) and end up with an unmaintainable mess. Componentization should be driven more by actual (not perceived) reusability, app organization, and substitution needs. You will end up with 5-10 XAML files for a menu (plus code-behind files) which has no value.

Discussing the actual problem: most collection controls require a specific child container type which contains required functionality for the parent. That container then can hold anything inside of it. This is true for `ListBox` with `ListBoxItem`, `Menu` with `MenuItem`, `MenuItem` also with `MenuItem`, `TreeView` with `TreeViewItem`, and others.

As such, you would need to derive from the container type. These are `TemplatedControl`s, so you don't get a coupled XAML file because they're lookless controls. Which means then you either: 1. create+assign `Header` and `Content` in code, plus use `StyleKeyOverride` to use the `MenuItem` `ControlTheme`. 2. create a `ControlTheme` (using `BasedOn`) for each so you can specify child content in XAML.

It's not a good place to be.

**Reply by Shadowblitz16** [reply]
> This is an antipattern that I call over-componentization. You try to create a lot of small components like you're writing React (you're not) and end up with an unmaintainable mess. Componentization should be driven more by actual (not perceived) reusability, app organization, and substitution needs. You will end up with 5-10 XAML files for a menu (plus code-behind files) which has no value.
> 
> Discussing the actual problem: most collection controls require a specific child container type which contains required functionality for the parent. That container then can hold anything inside of it. This is true for `ListBox` with `ListBoxItem`, `Menu` with `MenuItem`, `MenuItem` also with `MenuItem`, `TreeView` with `TreeViewItem`, and others.
> 
> As such, you would need to derive from the container type. These are `TemplatedControl`s, so you don't get a coupled XAML file because they're lookless controls. Which means then you either: 1. create+assign `Header` and `Content` in code, plus use `StyleKeyOverride` to use the `MenuItem` `ControlTheme`. 2. create a `ControlTheme` (using `BasedOn`) for each so you can specify child content in XAML.
> 
> It's not a good place to be.

That's too bad, Other gui toolkit could do this fine.
I find xaml horribly unreadable and I like to split my components up.

This should be fixed though because if it were declarative C# building the ui I could just split my Build  method into smaller parts.

**Reply by stevemonaco** [reply]
Other toolkits really don't. In every other GUI framework, you have instance wiring issues as your app grows large because you need to pull/push data or relations from several degrees of separation away. JS has a name for one such scenario: prop drilling. Avalonia is no different.

**Reply by Shadowblitz16** [reply]
> Other toolkits really don't. In every other GUI framework, you have instance wiring issues as your app grows large because you need to pull/push data or relations from several degrees of separation away. JS has a name for one such scenario: prop drilling. Avalonia is no different.

Yes they do. don't gas light me. It's literally just separating it into a different method, here is an example...
```cs
public class MyWindow : Form
{
    public MyWindow()
    {
        BuildRoot();
    }

     private void BuildRoot()
     {
           var mainMenu = BuildMenu();
           var panel          = BuildPanel();

           Menu = mainMenu;
           Children.Add(panel);
           
          
     }
     private Control BuildPanel()
     {
          var panel = new Panel(); 
          panel.Dock = Fill;
          return panel;
     }

     private Control BuildMenu()
     {
          var menu = new MainMenu();
          var help   = new MenuItem("Help");
          var about = new MenuItem("About");
          help.Items.Add(about);
          menu.Items.Add(help);
          return menu;
   
     }

}
```

How is that different then this?
```xaml
<Form 
...>
   <Menu/>
   <Panel/>
</Form>
```

```xaml
<Menu 
...>
    <MenuItem Header="_Help">
          <MenuItem Header="About"/>
    </MenuItem>
</Menu>
```

```xaml
<Panel
Dock="Fill"
...>
</Panel>
```

again look at what I said...
> This should be fixed though because if it were declarative C# building the ui I could just split my Build method into smaller parts.

**Reply by stevemonaco** [reply]
That's 5 controls which have no disconnected references to each other, have no external data, no bindings, no resources, no styling, no code-behind of their own, don't inherit from an existing control, and aren't custom controls. The controls are all owned by a single shared parent. It's a good example of a sample that looks clean but doesn't scale when complexity is introduced. It also runs into the same problem.

I use several strategies to cut down on XAML within a single file:
1. `DataTemplate`s are nearly equivalent to your simple factory methods. Can be defined externally. Lacks code-behind and parameter injection.
2. You can apply properties via styling. Can be defined externally.
3. You can separate into a `UserControl`.

Besides that, I use XAML Styler to autoformat XAML on save. Reading XAML is much easier when the formatting is consistent. Sometimes I will use comments between major sections instead of componentizing when there's no added value. I'm not building 500-1000 LoC XAML files, but I'm not building 5 LoC XAML files either.

**Reply by Shadowblitz16** [reply]
> That's 5 controls which have no disconnected references to each other, have no external data, no bindings, no resources, no styling, no code-behind of their own, don't inherit from an existing control, and aren't custom controls. The controls are all owned by a single shared parent. It's a good example of a sample that looks clean but doesn't scale when complexity is introduced. It also runs into the same problem.
> 
> I use several strategies to cut down on XAML within a single file:
> 
>     1. `DataTemplate`s are nearly equivalent to your simple factory methods. Can be defined externally. Lacks code-behind and parameter injection.
> 
>     2. You can apply properties via styling. Can be defined externally.
> 
>     3. You can separate into a `UserControl`.
> 
> 
> Besides that, I use XAML Styler to autoformat XAML on save. Reading XAML is much easier when the formatting is consistent. Sometimes I will use comments between major sections instead of componentizing when there's no added value. I'm not building 500-1000 LoC XAML files, but I'm not building 5 LoC XAML files either.

This may be 5 LOC in the xaml but the view model is larger. 
I am binding to ICommands which open dialogs and I am trying to separate my code into more readable chunks.

**Reply by stevemonaco** [reply]
If the ViewModel side is getting too large, consider writing Service Classes for the VM to call into. Which is just a fancy name for a utility class that performs a specific task (and is typically injected into your VM via IoC). This task would be offloaded from your VM, so now the VM command will only have a method call (and maybe some validation) instead of all of the logic, too. The maintenance cost of extra file sprawl is minimal and the benefits are immediate -- improved organization, separation of concerns, and testability.

You can also split VMs into multiple VMs. A XAML file can access multiple children VMs through the current one using the subproperty binding syntax. You don't need to make VMs and Views strictly 1:1.

**Comment by IanRawley** [comment]
What's the definition of EditorCanvasView in the code behind?  Is it derived from ScrollViewer?

**Reply by Shadowblitz16** [reply]
@IanRawley EditorCanvasView is just a panel.
I was trying to separate my view and view models into smaller parts.
The Menu is the main issue here since I was trying to separate it into separate view models per menu
